### PR TITLE
swift: 4.3.2 -> 5.0.1

### PR DESCRIPTION
**Pull request by mroi**
I updated Swift to version 5.0.1. This needed relatively minor changes, the biggest one is the addition of the `NIX_LDFLAGS_BEFORE` variable to help the internal clang that is being built during the Swift build process.

I also updated the patches and added two additional ones. One has been filed as an [upstream build issue](https://github.com/apple/swift-corelibs-libdispatch/pull/474), so it can be removed once this trickles down. The other causes the Swift build system not to create its own versions of libicu and libc++.

Some older build tweaks could be removed, because they no longer appear to have an effect.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dtzWill** [comment]
Awesome, thank you!!

This looks great! I can build it and seems to work as expected!

I think this can go in as-is-- I'll see about a few details and will make a follow-up PR if they turn out useful/desirable :).

**Comment by dtzWill** [comment]
Oh, I just saw the tests bit.  Should this wait for that to be sorted, then?
Wow, tests! That'd be great! :)

File: pkgs/development/compilers/swift/default.nix [hunk 1]
```diff
@@ -120,6 +120,9 @@ let
   builder = ''
     # gcc-6.4.0/include/c++/6.4.0/cstdlib:75:15: fatal error: 'stdlib.h' file not found
     NIX_CFLAGS_COMPILE="$( echo ${clang.default_cxx_stdlib_compile} ) $NIX_CFLAGS_COMPILE"
```
Comment (dtzWill) [inline pkgs/development/compilers/swift/default.nix]: Whhhyyyyyyy

1) Do we really not build with clang? :cry:
2) WTF putting CXXFLAGS in CFLAGS, yikes
3) WTF the mix of headers, compilers, runtime libraries-- do we even know what combination is used? Is it consistent throughout the various builds?!

..... And this isn't even something you're adding, so don't worry about it.  But... :(. D'oh.

**Comment by dtzWill** [comment]
Well, here's a few tweaks-- what do you think? If you like them feel free to pick them, or just say so and I can do it or make a PR or w/e.  Just wanted to share even if this PR isn't merged yet.

* Build w/default clang instead of old clang5: https://github.com/dtzWill/nixpkgs/commit/666029014d3252ad4673e86cbb13380c075c0a07
* Move most inputs expected to be on PATH to nativeBuildInputs: https://github.com/dtzWill/nixpkgs/commit/a365a5a3891bd83b21d61a9e5ff0314c02417f9a -- least confident about this, but seems to work.  And honestly this package is very likely to not work either way for cross, so it's mostly on principle I guess.
* Editor integration: https://github.com/dtzWill/nixpkgs/commit/f9332ff1ff356ab79da6d6fcae8882ad65501f41 -- installs at least vim and emacs bits from upstream.  May need some post-processing / plumbing to actually be used, I can at least tackle that for vim at some point.  Even so, better to have them available for use (even if they need some minor moving around), since building takes so long xD.

**Comment by mroi** [comment]
I would not wait for the tests to work out. These need some extra work and do not appear to be critical. The failures revolve around build problems that appear to be Nix-specific. They did not immediately indicate to me that the resulting swift compiler was broken.

Regarding your additional commits:
* I also tested building with default clang and would agree to use the stable version. Hardcoding clang_5 does not appear to have an obvious benefit.
* Regarding `nativeBuildInputs`: I think it conveys the intent of these dependencies better, so even if cross-builds are currently untested, I think this is a good idea to separate the dependencies we are confident about.
* Installing the editor integration sounds like a good idea as well. I wonder why the build presets only include it in the Darwin builds. Maybe this change should be propagated upstream?

In summary: I would vote for including all your proposed changes. If you want, I can include those in my pull request and retry the entire build at my end tomorrow.

File: pkgs/development/compilers/swift/default.nix (hunk 1 above)
Comment (mroi) [reply pkgs/development/compilers/swift/default.nix]: Yes, this is pretty gnarly. The way I see it, we do indeed use clang, but clang has gcc as a dependency and uses some of its compiler-shipping headers and startup files. The need for these extra flags however is a bit weird for me as well. I tried twice (but just briefly) to get rid of it, because I get bad vibes from it too, but so far failed to diagnose the actual problem. Nix’s clang should be able to find its stuff without help, but somehow it does not.

**Comment by mroi** [comment]
The way I see this, the original Nixpkgs repo already builds swift with default clang instead of hardcoding `clang_5`. So there must be something else going on.

The other two changes check out fine on my end. If you want, I can easily add them to this pull request.

**Comment by dtzWill** [comment]
On Mon, 06 May 2019 14:04:49 +0000 (UTC), Michael Roitzsch <notifications@github.com> wrote:
> The way I see this, the original Nixpkgs repo already builds swift with default clang instead of hardcoding `clang_5`. So there must be something else going on.

Whoops, sorry about that-- probably some local change I forgot :3.

> 
> The other two changes check out fine on my end. If you want, I can easily add them to this pull request.

Yes please, thanks!

**Comment by mroi** [comment]
I just pushed the additional commits.

**Comment by dtzWill** [comment]
Thank you for all your work on this!! LGTM!
